phoenix-core/src/main/java/org/apache/phoenix/jdbc/HighAvailabilityPolicy.java

```diff
@@ -0,0 +1,141 @@
+            LOG.info("Cluster {} becomes STANDBY in HA group {}, now close all its connections",
+                    zkUrl, haGroup.getGroupInfo());
+            ConnectionQueryServices cqs = null;
+            try {
+                cqs = PhoenixDriver.INSTANCE.getConnectionQueryServices(zkUrl,
+                         haGroup.getProperties());
+                cqs.closeAllConnections(new SQLExceptionInfo
+                        .Builder(SQLExceptionCode.HA_CLOSED_AFTER_FAILOVER)
+                        .setMessage("Phoenix connection got closed due to failover")
+                        .setHaGroupInfo(haGroup.getGroupInfo().toString()));
+                LOG.info("Closed all connections to cluster {} for HA group {}", zkUrl,
+                        haGroup.getGroupInfo());
+            } finally {
+                if (cqs != null) {
+                    // CQS is closed but it is not invalidated from global cache in PhoenixDriver
+                    // so that any new connection will get error instead of creating a new CQS
+                    LOG.info("Closing CQS after cluster '{}' becomes STANDBY", zkUrl);
+                    cqs.close();
+                    LOG.info("Successfully closed CQS after cluster '{}' becomes STANDBY", zkUrl);
+                }
+            }
+        }
+        private void transitActive(HighAvailabilityGroup haGroup, String zkUrl)
+                throws SQLException {
+            // Invalidate CQS cache if any that has been closed but has not been cleared
+            PhoenixDriver.INSTANCE.invalidateCache(zkUrl, haGroup.getProperties());
+        }
+    },
+
+    PARALLEL {
+        @Override
+        public Connection provide(HighAvailabilityGroup haGroup, Properties info)
+                throws SQLException {
+            List<Boolean> executorCapacities = PhoenixHAExecutorServiceProvider.hasCapacity(info);
+            if (executorCapacities.contains(Boolean.TRUE)) {
+                ParallelPhoenixContext context =
+                        new ParallelPhoenixContext(info, haGroup,
+                                PhoenixHAExecutorServiceProvider.get(info), executorCapacities);
+                return new ParallelPhoenixConnection(context);
+            } else {
+                // TODO: Once we have operation/primary wait timeout use the same
+                // Give regular connection or a failover connection?
+                LOG.warn("Falling back to single phoenix connection due to resource constraints");
+                GlobalClientMetrics.GLOBAL_HA_PARALLEL_CONNECTION_FALLBACK_COUNTER.increment();
+                return haGroup.connectActive(info);
+            }
+        }
+        @Override
+        void transitClusterRole(HighAvailabilityGroup haGroup, ClusterRoleRecord oldRecord,
+                ClusterRoleRecord newRecord) {
+            LOG.info("Cluster role changed for parallel HA policy.");
```

Comment:
We don't expect anything to change so it would be good to have a log that something tried to change it.  Parallel does not differentiate between any state.